### PR TITLE
PR#6217 Improve perf. of functional record update

### PR DESCRIPTION
**Pull request by OlivierNicole**
See [Mantis issue #6217](http://caml.inria.fr/mantis/view.php?id=6217). Benchmarking with `Core_bench` on a 2.5 GHz 64-bit Intel i5-2450M showed that fieldwise copy is always faster while the number of fields is smaller than `Max_young_wosize` (that is, 256); beyond that point the "duplication and modification" method becomes much faster.

Benchmark details:
[rfu_benchmarks.pdf](https://github.com/ocaml/ocaml/files/218048/rfu_benchmarks.pdf)
CSV results:
[rfu_bench.zip](https://github.com/ocaml/ocaml/files/218054/rfu_bench.zip)

### CODE REVIEW

File: bytecomp/translcore.ml [hunk 1]
```diff
@@ -1235,7 +1235,8 @@ and transl_setinstvar self var expr =
 and transl_record env all_labels repres lbl_expr_list opt_init_expr =
   let size = Array.length all_labels in
   (* Determine if there are "enough" new fields *)
-  if 3 + 2 * List.length lbl_expr_list >= size
+  let no_init = match opt_init_expr with None -> true | _ -> false in
+  if no_init || size > 255
```
Comment (braibant) [inline bytecomp/translcore.ml]: Should this be  `Config.max_young_wosize` rather than "255"?

Comment (OlivierNicole) [inline bytecomp/translcore.ml]: Indeed, thank you.

**Comment by yminsky** [comment]
I'd be interested in seeing the benchmarks, if that's possible.  Perhaps just attach the benchmark details to the PR?

**Comment by OlivierNicole** [comment]
Hello,
I just attached a PDF with details and graphs.

**Comment by gasche** [comment]
Thanks for the detailed measurements. I merged the pull request in trunk.

File: bytecomp/translcore.ml [hunk 2]
```diff
@@ -1234,8 +1234,10 @@ and transl_setinstvar self var expr =
 
 and transl_record env all_labels repres lbl_expr_list opt_init_expr =
   let size = Array.length all_labels in
-  (* Determine if there are "enough" new fields *)
-  if 3 + 2 * List.length lbl_expr_list >= size
+  (* Determine if there are "enough" fields (only relevant if this is a
+     functional-style record update *)
+  let no_init = match opt_init_expr with None -> true | _ -> false in
+  if no_init || size >= Config.max_young_wosize
```
Comment (chambart) [inline bytecomp/translcore.ml:1240]: The direction of this test is wrong: should be size <= Config.max_young_wosize

Comment (OlivierNicole) [inline bytecomp/translcore.ml:1240]: Oops… sorry about that. Should I make a new pull request?

I believe "<" would be better since the `caml_modify` method is already faster with a 256 fields.

Comment (chambart) [inline bytecomp/translcore.ml:1240]: A fix is part of https://github.com/ocaml/ocaml/pull/516

Comment (gasche) [inline bytecomp/translcore.ml:1240]: @chambart good catch! Sorry, I did review this but these conditions are trick.

I pushed a fix in 755427318e4ffa1c23b54ab10f11d66e7c4fcf58 , I think it's better to separate concerns and leave your own PR about one thing.

Comment (bobzhang) [reply bytecomp/translcore.ml:1240]: this will cause code bloat in pathological cases

Comment (lpw25) [reply bytecomp/translcore.ml:1240]: What cases do you have in mind? If the size is of the record is larger than 256 then it uses the "duplicate and modify" approach which requires less code.
